// abstract/static/lib-cov/utility.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['utility.js']) {
  _$jscoverage['utility.js'] = [];
  _$jscoverage['utility.js'][14] = 0;
  _$jscoverage['utility.js'][15] = 0;
  _$jscoverage['utility.js'][18] = 0;
  _$jscoverage['utility.js'][32] = 0;
  _$jscoverage['utility.js'][33] = 0;
  _$jscoverage['utility.js'][36] = 0;
  _$jscoverage['utility.js'][37] = 0;
  _$jscoverage['utility.js'][40] = 0;
  _$jscoverage['utility.js'][42] = 0;
  _$jscoverage['utility.js'][52] = 0;
  _$jscoverage['utility.js'][53] = 0;
  _$jscoverage['utility.js'][55] = 0;
  _$jscoverage['utility.js'][57] = 0;
  _$jscoverage['utility.js'][58] = 0;
  _$jscoverage['utility.js'][61] = 0;
  _$jscoverage['utility.js'][62] = 0;
  _$jscoverage['utility.js'][64] = 0;
  _$jscoverage['utility.js'][87] = 0;
  _$jscoverage['utility.js'][89] = 0;
  _$jscoverage['utility.js'][90] = 0;
  _$jscoverage['utility.js'][91] = 0;
  _$jscoverage['utility.js'][92] = 0;
  _$jscoverage['utility.js'][95] = 0;
  _$jscoverage['utility.js'][97] = 0;
  _$jscoverage['utility.js'][99] = 0;
  _$jscoverage['utility.js'][101] = 0;
  _$jscoverage['utility.js'][116] = 0;
  _$jscoverage['utility.js'][118] = 0;
  _$jscoverage['utility.js'][119] = 0;
  _$jscoverage['utility.js'][120] = 0;
  _$jscoverage['utility.js'][123] = 0;
  _$jscoverage['utility.js'][125] = 0;
  _$jscoverage['utility.js'][127] = 0;
  _$jscoverage['utility.js'][140] = 0;
  _$jscoverage['utility.js'][141] = 0;
  _$jscoverage['utility.js'][142] = 0;
  _$jscoverage['utility.js'][143] = 0;
  _$jscoverage['utility.js'][145] = 0;
  _$jscoverage['utility.js'][148] = 0;
  _$jscoverage['utility.js'][149] = 0;
  _$jscoverage['utility.js'][177] = 0;
  _$jscoverage['utility.js'][178] = 0;
  _$jscoverage['utility.js'][180] = 0;
  _$jscoverage['utility.js'][183] = 0;
  _$jscoverage['utility.js'][185] = 0;
  _$jscoverage['utility.js'][187] = 0;
  _$jscoverage['utility.js'][192] = 0;
  _$jscoverage['utility.js'][195] = 0;
  _$jscoverage['utility.js'][196] = 0;
  _$jscoverage['utility.js'][199] = 0;
  _$jscoverage['utility.js'][200] = 0;
  _$jscoverage['utility.js'][203] = 0;
  _$jscoverage['utility.js'][204] = 0;
  _$jscoverage['utility.js'][218] = 0;
  _$jscoverage['utility.js'][219] = 0;
  _$jscoverage['utility.js'][245] = 0;
  _$jscoverage['utility.js'][246] = 0;
  _$jscoverage['utility.js'][249] = 0;
  _$jscoverage['utility.js'][273] = 0;
  _$jscoverage['utility.js'][274] = 0;
  _$jscoverage['utility.js'][275] = 0;
  _$jscoverage['utility.js'][280] = 0;
  _$jscoverage['utility.js'][281] = 0;
  _$jscoverage['utility.js'][284] = 0;
  _$jscoverage['utility.js'][302] = 0;
  _$jscoverage['utility.js'][303] = 0;
  _$jscoverage['utility.js'][306] = 0;
  _$jscoverage['utility.js'][307] = 0;
  _$jscoverage['utility.js'][310] = 0;
  _$jscoverage['utility.js'][311] = 0;
  _$jscoverage['utility.js'][314] = 0;
  _$jscoverage['utility.js'][316] = 0;
  _$jscoverage['utility.js'][318] = 0;
  _$jscoverage['utility.js'][319] = 0;
  _$jscoverage['utility.js'][324] = 0;
  _$jscoverage['utility.js'][363] = 0;
  _$jscoverage['utility.js'][364] = 0;
  _$jscoverage['utility.js'][368] = 0;
  _$jscoverage['utility.js'][371] = 0;
  _$jscoverage['utility.js'][393] = 0;
  _$jscoverage['utility.js'][394] = 0;
  _$jscoverage['utility.js'][395] = 0;
  _$jscoverage['utility.js'][397] = 0;
  _$jscoverage['utility.js'][398] = 0;
  _$jscoverage['utility.js'][399] = 0;
  _$jscoverage['utility.js'][400] = 0;
  _$jscoverage['utility.js'][401] = 0;
  _$jscoverage['utility.js'][405] = 0;
  _$jscoverage['utility.js'][406] = 0;
  _$jscoverage['utility.js'][409] = 0;
  _$jscoverage['utility.js'][411] = 0;
  _$jscoverage['utility.js'][414] = 0;
  _$jscoverage['utility.js'][417] = 0;
  _$jscoverage['utility.js'][420] = 0;
  _$jscoverage['utility.js'][444] = 0;
  _$jscoverage['utility.js'][445] = 0;
  _$jscoverage['utility.js'][448] = 0;
  _$jscoverage['utility.js'][449] = 0;
  _$jscoverage['utility.js'][452] = 0;
  _$jscoverage['utility.js'][454] = 0;
  _$jscoverage['utility.js'][471] = 0;
  _$jscoverage['utility.js'][472] = 0;
  _$jscoverage['utility.js'][475] = 0;
  _$jscoverage['utility.js'][479] = 0;
  _$jscoverage['utility.js'][480] = 0;
  _$jscoverage['utility.js'][481] = 0;
  _$jscoverage['utility.js'][485] = 0;
  _$jscoverage['utility.js'][515] = 0;
  _$jscoverage['utility.js'][516] = 0;
  _$jscoverage['utility.js'][517] = 0;
  _$jscoverage['utility.js'][518] = 0;
  _$jscoverage['utility.js'][519] = 0;
  _$jscoverage['utility.js'][521] = 0;
  _$jscoverage['utility.js'][522] = 0;
  _$jscoverage['utility.js'][523] = 0;
  _$jscoverage['utility.js'][525] = 0;
  _$jscoverage['utility.js'][526] = 0;
  _$jscoverage['utility.js'][529] = 0;
  _$jscoverage['utility.js'][533] = 0;
  _$jscoverage['utility.js'][550] = 0;
  _$jscoverage['utility.js'][551] = 0;
  _$jscoverage['utility.js'][553] = 0;
  _$jscoverage['utility.js'][555] = 0;
  _$jscoverage['utility.js'][556] = 0;
  _$jscoverage['utility.js'][562] = 0;
  _$jscoverage['utility.js'][566] = 0;
  _$jscoverage['utility.js'][582] = 0;
  _$jscoverage['utility.js'][583] = 0;
  _$jscoverage['utility.js'][587] = 0;
  _$jscoverage['utility.js'][588] = 0;
  _$jscoverage['utility.js'][591] = 0;
  _$jscoverage['utility.js'][604] = 0;
  _$jscoverage['utility.js'][608] = 0;
  _$jscoverage['utility.js'][617] = 0;
  _$jscoverage['utility.js'][619] = 0;
  _$jscoverage['utility.js'][621] = 0;
  _$jscoverage['utility.js'][623] = 0;
  _$jscoverage['utility.js'][624] = 0;
  _$jscoverage['utility.js'][627] = 0;
  _$jscoverage['utility.js'][629] = 0;
  _$jscoverage['utility.js'][631] = 0;
  _$jscoverage['utility.js'][635] = 0;
  _$jscoverage['utility.js'][639] = 0;
  _$jscoverage['utility.js'][654] = 0;
  _$jscoverage['utility.js'][656] = 0;
  _$jscoverage['utility.js'][678] = 0;
  _$jscoverage['utility.js'][679] = 0;
  _$jscoverage['utility.js'][681] = 0;
  _$jscoverage['utility.js'][682] = 0;
  _$jscoverage['utility.js'][684] = 0;
  _$jscoverage['utility.js'][686] = 0;
  _$jscoverage['utility.js'][687] = 0;
  _$jscoverage['utility.js'][689] = 0;
  _$jscoverage['utility.js'][711] = 0;
  _$jscoverage['utility.js'][712] = 0;
  _$jscoverage['utility.js'][715] = 0;
  _$jscoverage['utility.js'][716] = 0;
  _$jscoverage['utility.js'][719] = 0;
  _$jscoverage['utility.js'][734] = 0;
  _$jscoverage['utility.js'][735] = 0;
  _$jscoverage['utility.js'][736] = 0;
  _$jscoverage['utility.js'][738] = 0;
  _$jscoverage['utility.js'][762] = 0;
  _$jscoverage['utility.js'][763] = 0;
  _$jscoverage['utility.js'][764] = 0;
  _$jscoverage['utility.js'][770] = 0;
  _$jscoverage['utility.js'][772] = 0;
  _$jscoverage['utility.js'][773] = 0;
  _$jscoverage['utility.js'][776] = 0;
  _$jscoverage['utility.js'][778] = 0;
  _$jscoverage['utility.js'][781] = 0;
  _$jscoverage['utility.js'][798] = 0;
  _$jscoverage['utility.js'][799] = 0;
  _$jscoverage['utility.js'][802] = 0;
  _$jscoverage['utility.js'][803] = 0;
  _$jscoverage['utility.js'][806] = 0;
  _$jscoverage['utility.js'][820] = 0;
  _$jscoverage['utility.js'][821] = 0;
  _$jscoverage['utility.js'][824] = 0;
  _$jscoverage['utility.js'][825] = 0;
  _$jscoverage['utility.js'][826] = 0;
  _$jscoverage['utility.js'][827] = 0;
  _$jscoverage['utility.js'][830] = 0;
  _$jscoverage['utility.js'][868] = 0;
  _$jscoverage['utility.js'][871] = 0;
  _$jscoverage['utility.js'][872] = 0;
  _$jscoverage['utility.js'][875] = 0;
  _$jscoverage['utility.js'][877] = 0;
  _$jscoverage['utility.js'][879] = 0;
  _$jscoverage['utility.js'][881] = 0;
  _$jscoverage['utility.js'][882] = 0;
  _$jscoverage['utility.js'][884] = 0;
  _$jscoverage['utility.js'][889] = 0;
  _$jscoverage['utility.js'][891] = 0;
}
_$jscoverage['utility.js'][14]++;
function Utility() {
  _$jscoverage['utility.js'][15]++;
  return Object.keys(arguments)? require("lodash").pick.apply(null, [Utility, Array.prototype.slice.call(arguments)]): Utility;
}
_$jscoverage['utility.js'][18]++;
Object.defineProperties(module.exports = Utility, {json: {value: {parse: (function (json) {
  _$jscoverage['utility.js'][32]++;
  if (Buffer && json instanceof Buffer) {
    _$jscoverage['utility.js'][33]++;
    json = json.toString();
  }
  _$jscoverage['utility.js'][36]++;
  var parsed = false;
  _$jscoverage['utility.js'][37]++;
  try {
    _$jscoverage['utility.js'][37]++;
    parsed = JSON.parse(json);
  }
  catch (e) {
    _$jscoverage['utility.js'][37]++;
    parsed = false;
  }
  _$jscoverage['utility.js'][40]++;
  if (! parsed && "string" === typeof json) {
  }
  _$jscoverage['utility.js'][42]++;
  return parsed? parsed: json;
})}, configurable: false, enumerable: true, writable: true}, realpath: {value: (function realpath() {
  _$jscoverage['utility.js'][52]++;
  var join_path = require("path").join;
  _$jscoverage['utility.js'][53]++;
  var fs = require("fs");
  _$jscoverage['utility.js'][55]++;
  var _path = arguments.length > 1? join_path.apply({}, arguments): arguments[0];
  _$jscoverage['utility.js'][57]++;
  try {
    _$jscoverage['utility.js'][58]++;
    return fs.realpathSync(_path);
  }
  catch (error) {
    _$jscoverage['utility.js'][61]++;
    try {
      _$jscoverage['utility.js'][62]++;
      return require.resolve(_path);
    }
    catch (error) {
      _$jscoverage['utility.js'][64]++;
      return null;
    }
  }
}), configurable: false, enumerable: true, writable: true}, console: {value: {json: (function json(data, depth) {
  _$jscoverage['utility.js'][87]++;
  try {
    _$jscoverage['utility.js'][89]++;
    if ("string" === typeof data && ("object" === typeof depth || "function" === typeof depth)) {
      _$jscoverage['utility.js'][90]++;
      console.log("\n" + "===============" + data + " Properties===============");
      _$jscoverage['utility.js'][91]++;
      data = depth;
      _$jscoverage['utility.js'][92]++;
      depth = arguments[2] || 1;
    }
    _$jscoverage['utility.js'][95]++;
    var output = require("util").inspect(data, false, "number" === typeof depth? (depth - 1): 1, true);
    _$jscoverage['utility.js'][97]++;
    console.log(output);
  }
  catch (error) {
    _$jscoverage['utility.js'][99]++;
    console.error("Error with console.json()", error);
  }
  _$jscoverage['utility.js'][101]++;
  return arguments[0];
}), method: (function method(data, object) {
  _$jscoverage['utility.js'][116]++;
  try {
    _$jscoverage['utility.js'][118]++;
    if ("string" === typeof data && ("object" === typeof object || "function" === typeof object)) {
      _$jscoverage['utility.js'][119]++;
      console.log("\n" + "===============" + data + " Methods===============");
      _$jscoverage['utility.js'][120]++;
      data = object;
    }
    _$jscoverage['utility.js'][123]++;
    module.exports.json(_.methods(data));
  }
  catch (error) {
    _$jscoverage['utility.js'][125]++;
    console.error("Error with console.methods()", error);
  }
  _$jscoverage['utility.js'][127]++;
  return arguments[0];
}), keys: (function keys(data) {
  _$jscoverage['utility.js'][140]++;
  function Iterate(target) {
    _$jscoverage['utility.js'][141]++;
    var result = [];
    _$jscoverage['utility.js'][142]++;
    for (var key in target) {
      _$jscoverage['utility.js'][143]++;
      result.push(key);
}
    _$jscoverage['utility.js'][145]++;
    return result;
}
  _$jscoverage['utility.js'][148]++;
  console.log([Object.keys(data) + Object.getOwnPropertyNames(data) + Iterate(data)]);
  _$jscoverage['utility.js'][149]++;
  return data;
}), table: (function table(data) {
  _$jscoverage['utility.js'][177]++;
  try {
    _$jscoverage['utility.js'][178]++;
    var table = require("cli-table");
  }
  catch (error) {
    _$jscoverage['utility.js'][180]++;
    return console.log(data);
  }
  _$jscoverage['utility.js'][183]++;
  var rows = [];
  _$jscoverage['utility.js'][185]++;
  if (_.isObject(data) && data.rows) {
    _$jscoverage['utility.js'][187]++;
    table = new table({"head": data.head || [], "colWidths": data.colWidths || []});
    _$jscoverage['utility.js'][192]++;
    rows = data.rows;
  }
  else {
    _$jscoverage['utility.js'][195]++;
    table = new table();
    _$jscoverage['utility.js'][196]++;
    rows = data;
  }
  _$jscoverage['utility.js'][199]++;
  _.each(rows, (function (row) {
  _$jscoverage['utility.js'][200]++;
  if (! _.isEmpty(row)) {
    _$jscoverage['utility.js'][200]++;
    table.push(row);
  }
}));
  _$jscoverage['utility.js'][203]++;
  try {
    _$jscoverage['utility.js'][204]++;
    console.log(table.toString());
  }
  catch (error) {
  }
}), report: (function report() {
  _$jscoverage['utility.js'][218]++;
  module.exports.console.json.apply(module.exports, arguments);
  _$jscoverage['utility.js'][219]++;
  module.exports.console.methods.apply(module.exports, arguments);
})}, configurable: false, enumerable: true, writable: true}, merge: {value: (function merge(a, b) {
  _$jscoverage['utility.js'][245]++;
  if (a && b) {
    _$jscoverage['utility.js'][246]++;
    for (var key in b) {
      _$jscoverage['utility.js'][246]++;
      a[key] = b[key];
}
  }
  _$jscoverage['utility.js'][249]++;
  return a;
}), configurable: false, enumerable: true, writable: true}, embed: {value: (function embed(target) {
  _$jscoverage['utility.js'][273]++;
  if ("object" === typeof target || target.keys) {
    _$jscoverage['utility.js'][274]++;
    for (var key in target) {
      _$jscoverage['utility.js'][275]++;
      this[key] = target[key];
}
  }
  _$jscoverage['utility.js'][280]++;
  if ("function" === typeof target) {
    _$jscoverage['utility.js'][281]++;
    return Utility.embed(this);
  }
  _$jscoverage['utility.js'][284]++;
  return this;
}), configurable: false, enumerable: true, writable: true}, regroup: {value: (function regroup(object, group_key, keys) {
  _$jscoverage['utility.js'][302]++;
  if (_.isObject(object)) {
    _$jscoverage['utility.js'][303]++;
    object = _.flatten(_.values(object));
  }
  _$jscoverage['utility.js'][306]++;
  function Group(item) {
    _$jscoverage['utility.js'][307]++;
    return item[group_key];
}
  _$jscoverage['utility.js'][310]++;
  if (group_key) {
    _$jscoverage['utility.js'][311]++;
    object = _.groupBy(object, Group);
  }
  _$jscoverage['utility.js'][314]++;
  if (! _.isEmpty(keys)) {
    _$jscoverage['utility.js'][316]++;
    keys = _.isArray(keys)? keys: [keys];
    _$jscoverage['utility.js'][318]++;
    _.each(object, (function (items, key) {
  _$jscoverage['utility.js'][319]++;
  object[key] = _.map(items, (function (item) {
  _$jscoverage['utility.js'][319]++;
  return _.pick(item, keys);
}));
}));
  }
  _$jscoverage['utility.js'][324]++;
  return object;
}), configurable: false, enumerable: true, writable: true}, noop: {value: (function noop() {
}), configurable: false, enumerable: true, writable: true}, apply: {value: require("async").apply, configurable: true, enumerable: true, writable: true}, defaults: {value: (function defaults(target, defaults) {
  _$jscoverage['utility.js'][363]++;
  if (! target && ! defaults) {
    _$jscoverage['utility.js'][364]++;
    return ({});
  }
  _$jscoverage['utility.js'][368]++;
  target = Object.keys(target || {}).length? target: {};
  _$jscoverage['utility.js'][371]++;
  return require("lodash").defaults.apply({}, arguments);
}), configurable: true, enumerable: true, writable: true}, watch: {value: (function watch(target, prop, handler) {
  _$jscoverage['utility.js'][393]++;
  var oldval = target[prop];
  _$jscoverage['utility.js'][394]++;
  var newval = oldval;
  _$jscoverage['utility.js'][395]++;
  var description = Object.getOwnPropertyDescriptor(target, prop) || {};
  _$jscoverage['utility.js'][397]++;
  switch (typeof target[prop]) {
  case "function":
    _$jscoverage['utility.js'][398]++;
    break;
  case "object":
    _$jscoverage['utility.js'][399]++;
    break;
  case "string":
    _$jscoverage['utility.js'][400]++;
    break;
  case "number":
    _$jscoverage['utility.js'][401]++;
    break;
  }
  _$jscoverage['utility.js'][405]++;
  if (description.configurable === false || ! (delete target[prop])) {
    _$jscoverage['utility.js'][406]++;
    return target;
  }
  _$jscoverage['utility.js'][409]++;
  return Object.defineProperty(target, prop, {get: (function get() {
  _$jscoverage['utility.js'][411]++;
  return newval;
}), set: (function set(val) {
  _$jscoverage['utility.js'][414]++;
  oldval = newval;
  _$jscoverage['utility.js'][417]++;
  newval = val;
  _$jscoverage['utility.js'][420]++;
  return newval = handler.call(target, prop, oldval, val);
}), enumerable: description.enumerable, configurable: true});
}), configurable: true, enumerable: true, writable: true}, unwatch: {value: (function unwatch(prop) {
  _$jscoverage['utility.js'][444]++;
  var val = this[prop];
  _$jscoverage['utility.js'][445]++;
  var description = Object.getOwnPropertyDescriptor(this, prop);
  _$jscoverage['utility.js'][448]++;
  if (description.configurable === false || ! (delete this[prop])) {
    _$jscoverage['utility.js'][449]++;
    return this;
  }
  _$jscoverage['utility.js'][452]++;
  this[prop] = val;
  _$jscoverage['utility.js'][454]++;
  return this;
}), configurable: true, enumerable: true, writable: true}, query: {value: (function query(obj, str) {
  _$jscoverage['utility.js'][471]++;
  if (! str) {
    _$jscoverage['utility.js'][472]++;
    return obj;
  }
  _$jscoverage['utility.js'][475]++;
  if ("object" === typeof str) {
  }
  _$jscoverage['utility.js'][479]++;
  try {
    _$jscoverage['utility.js'][480]++;
    return str.split(".").reduce((function (o, x) {
  _$jscoverage['utility.js'][481]++;
  return o[x];
}), obj);
  }
  catch (error) {
    _$jscoverage['utility.js'][485]++;
    return null;
  }
}), configurable: true, enumerable: true, writable: true}, unwrap: {value: (function unwrap(string, value, hash, seperator) {
  _$jscoverage['utility.js'][515]++;
  if (hash == null) {
    _$jscoverage['utility.js'][515]++;
    hash = {};
  }
  _$jscoverage['utility.js'][516]++;
  if (seperator == null) {
    _$jscoverage['utility.js'][516]++;
    seperator = ".";
  }
  _$jscoverage['utility.js'][517]++;
  var parts = string.split(seperator);
  _$jscoverage['utility.js'][518]++;
  var refHash = hash;
  _$jscoverage['utility.js'][519]++;
  var depth = 0;
  _$jscoverage['utility.js'][521]++;
  parts.forEach((function (part) {
  _$jscoverage['utility.js'][522]++;
  if (depth == parts.length - 1) {
    _$jscoverage['utility.js'][523]++;
    refHash[part] = value;
  }
  else {
    _$jscoverage['utility.js'][525]++;
    if (refHash[part] == null) {
      _$jscoverage['utility.js'][525]++;
      refHash[part] = {};
    }
    _$jscoverage['utility.js'][526]++;
    refHash = refHash[part];
  }
  _$jscoverage['utility.js'][529]++;
  depth++;
}));
  _$jscoverage['utility.js'][533]++;
  return hash;
}), configurable: true, enumerable: true, writable: true}, mixin: {value: (function mixin(source, target) {
  _$jscoverage['utility.js'][550]++;
  source = this.prototype || arguments[0] || {};
  _$jscoverage['utility.js'][551]++;
  target = arguments.length == 2? arguments[1]: arguments[1] || {};
  _$jscoverage['utility.js'][553]++;
  for (var key in source) {
    _$jscoverage['utility.js'][555]++;
    try {
      _$jscoverage['utility.js'][556]++;
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key) || {value: source[key], enumerable: false, writable: true, configurable: true});
    }
    catch (error) {
      _$jscoverage['utility.js'][562]++;
      console.error(error);
    }
}
  _$jscoverage['utility.js'][566]++;
  return target;
}), configurable: true, writable: true, enumerable: false}, inherit: {value: (function inherit(target, source) {
  _$jscoverage['utility.js'][582]++;
  target = target || {};
  _$jscoverage['utility.js'][583]++;
  source = source || {};
  _$jscoverage['utility.js'][587]++;
  Object.getOwnPropertyNames(source).forEach((function (key) {
  _$jscoverage['utility.js'][588]++;
  if (! target.hasOwnProperty(key)) {
    _$jscoverage['utility.js'][588]++;
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }
}));
  _$jscoverage['utility.js'][591]++;
  return target;
}), enumerable: true}, flatten: {value: (function flatten(data, options) {
  _$jscoverage['utility.js'][604]++;
  options = Utility.defaults(options, {delimiter: 1});
  _$jscoverage['utility.js'][608]++;
  return Array.prototype.concat(data).join(options.delimiter).toLowerCase();
}), enumerable: true}, flatten_obj: {value: (function flatten_obj(obj) {
  _$jscoverage['utility.js'][617]++;
  var list = {};
  _$jscoverage['utility.js'][619]++;
  (function callee(o, r) {
  _$jscoverage['utility.js'][621]++;
  r = r || "";
  _$jscoverage['utility.js'][623]++;
  if (typeof o != "object") {
    _$jscoverage['utility.js'][624]++;
    return true;
  }
  _$jscoverage['utility.js'][627]++;
  for (var c in o) {
    _$jscoverage['utility.js'][629]++;
    if (callee(o[c], r + "." + c)) {
      _$jscoverage['utility.js'][631]++;
      list[r.substring(1) + "." + c] = o[c];
    }
}
  _$jscoverage['utility.js'][635]++;
  return false;
})(obj);
  _$jscoverage['utility.js'][639]++;
  return list;
}), enumerable: true}, inherit_full: {value: (function inherit_prototype(target, constructor) {
  _$jscoverage['utility.js'][654]++;
  target.super_ = constructor;
  _$jscoverage['utility.js'][656]++;
  target.prototype = Object.create(constructor.prototype, {constructor: {value: constructor, enumerable: false, writable: true, configurable: true}});
}), configurable: true, enumerable: true, writable: true}, constructors: {value: (function constructors(context, options) {
  _$jscoverage['utility.js'][678]++;
  context = context || {};
  _$jscoverage['utility.js'][679]++;
  options = options || {};
  _$jscoverage['utility.js'][681]++;
  var matches = {};
  _$jscoverage['utility.js'][682]++;
  var path = [];
  _$jscoverage['utility.js'][684]++;
  do {
    _$jscoverage['utility.js'][686]++;
    if (context.constructor) {
      _$jscoverage['utility.js'][687]++;
      path.push(context.constructor.name);
      _$jscoverage['utility.js'][689]++;
      Object.defineProperty(matches, context.constructor.name, {enumerable: true, writable: true, value: {name: context.constructor.name, properties: {constructor: {type: typeof context.constructor, properties: Object.getOwnPropertyNames(context.constructor)}, prototype: {type: typeof context.constructor.prototype, properties: Object.keys(context.constructor.prototype)}}}});
    }
}
  while (context = Object.getPrototypeOf(context));
  _$jscoverage['utility.js'][711]++;
  if (options.format === "schema") {
    _$jscoverage['utility.js'][712]++;
    return matches || {};
  }
  _$jscoverage['utility.js'][715]++;
  if (options.delimiter) {
    _$jscoverage['utility.js'][716]++;
    return path.join(options.delimiter);
  }
  _$jscoverage['utility.js'][719]++;
  return path;
}), configurable: true, enumerable: true, writable: true}, unique: {value: (function unique(array) {
  _$jscoverage['utility.js'][734]++;
  var b = [];
  _$jscoverage['utility.js'][735]++;
  for (var i = 0; i < array.length; i++) {
    _$jscoverage['utility.js'][736]++;
    if (b.indexOf(array[i]) == -1) {
      _$jscoverage['utility.js'][736]++;
      b.push(array[i]);
    }
}
  _$jscoverage['utility.js'][738]++;
  return b;
}), configurable: true, enumerable: true, writable: true}, hash: {value: (function hash(obj) {
  _$jscoverage['utility.js'][762]++;
  var result;
  _$jscoverage['utility.js'][763]++;
  var string = JSON.stringify(arguments[0] || {});
  _$jscoverage['utility.js'][764]++;
  var options = Utility.defaults(arguments[1], {type: "md5", silent_fail: true, random: false});
  _$jscoverage['utility.js'][770]++;
  try {
    _$jscoverage['utility.js'][772]++;
    if (options.random) {
      _$jscoverage['utility.js'][773]++;
      string = Math.random().toString();
    }
    _$jscoverage['utility.js'][776]++;
    result = require("crypto").createHash(options.type.toLowerCase()).update(string).digest("hex");
  }
  catch (error) {
    _$jscoverage['utility.js'][778]++;
    result = options.silent_fail? "": error;
  }
  _$jscoverage['utility.js'][781]++;
  return result;
}), configurable: true, enumerable: true, writable: true}, auto: {value: require("async").auto}, queue: {value: require("async").queue}, times: {value: require("async").times}, extend: {value: require("lodash").extend}, values: {value: require("lodash").values}, map: {value: require("lodash").map}, each: {value: require("lodash").each}, reduce: {value: (function reduce(arr, iterator, memo) {
  _$jscoverage['utility.js'][798]++;
  if (arr.reduce) {
    _$jscoverage['utility.js'][799]++;
    return arr.reduce(iterator, memo);
  }
  _$jscoverage['utility.js'][802]++;
  _each(arr, (function (x, i, a) {
  _$jscoverage['utility.js'][803]++;
  memo = iterator(memo, x, i, a);
}));
  _$jscoverage['utility.js'][806]++;
  return memo;
}), configurable: false, enumerable: true, writable: true}, keys: {value: (function keys(obj) {
  _$jscoverage['utility.js'][820]++;
  if (Object.keys) {
    _$jscoverage['utility.js'][821]++;
    return Object.keys(obj);
  }
  _$jscoverage['utility.js'][824]++;
  var keys = [];
  _$jscoverage['utility.js'][825]++;
  for (var k in obj) {
    _$jscoverage['utility.js'][826]++;
    if (obj.hasOwnProperty(k)) {
      _$jscoverage['utility.js'][827]++;
      keys.push(k);
    }
}
  _$jscoverage['utility.js'][830]++;
  return keys;
}), configurable: false, enumerable: true, writable: true}, toArray: {value: require("lodash").toArray}, where: {value: require("lodash").where}, if: {value: {"PlainObject": require("lodash").isPlainObject, "Function": require("lodash").isFunction, "Object": require("lodash").isObject, "String": require("lodash").isString}, configurable: false, enumerable: true, writable: true}, log: {value: Object.defineProperties((function () {
  _$jscoverage['utility.js'][868]++;
  var output = {};
  _$jscoverage['utility.js'][871]++;
  if (this.event) {
    _$jscoverage['utility.js'][872]++;
    output.event = this.event;
  }
  _$jscoverage['utility.js'][875]++;
  Utility.each(arguments, (function (item, key) {
  _$jscoverage['utility.js'][877]++;
  if (Utility["if"].PlainObject(item)) {
  }
  _$jscoverage['utility.js'][879]++;
  output.format = typeof item;
  _$jscoverage['utility.js'][881]++;
  if (key === 0) {
    _$jscoverage['utility.js'][882]++;
    output.data = item;
  }
  else {
    _$jscoverage['utility.js'][884]++;
    output[key] = item;
  }
}));
  _$jscoverage['utility.js'][889]++;
  console.log(require("util").inspect(output, Utility.log.config));
  _$jscoverage['utility.js'][891]++;
  return this;
}), {"config": {value: {"showHidden": true, "depth": 2, "colors": true}, "writable": true}}), enumerable: true}});
_$jscoverage['utility.js'].source = ["/**"," * Helper Utility for Abstract"," *"," * @example"," *"," *    // Select specific methods to load"," *    var my_tools = Abstract.utility( 'if', 'extend', 'flatten', 'unwatch', 'watch' );"," *"," * @for abstract"," * @submodule abstract-utility"," * @author potanin@UD"," * @date 6/17/13"," */","function Utility() {","  return Object.keys( arguments ) ? require( 'lodash' ).pick.apply( null, [ Utility, Array.prototype.slice.call( arguments ) ] ) : Utility;","}","","Object.defineProperties( module.exports = Utility, {","  json: {","    value: {","      /**","       * Safely parse JSON","       *","       * @method parse","       * @for Json","       *","       * @param json","       * @return {parsed|json}","       */","      parse: function( json ) {","","        if( Buffer &amp;&amp; json instanceof Buffer ) {","          json = json.toString()","        }","","        var parsed = false;","        try { parsed = JSON.parse( json ); } catch (e) { parsed = false; }","","        // Perhaps try to strip some line braks or special characters..","        if( !parsed &amp;&amp; 'string' === typeof json ) {}","","        return parsed ? parsed : json;","      }","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  realpath: {","    value: function realpath() {","","      var join_path = require( 'path' ).join;","      var fs = require( 'fs' );","","      var _path = arguments.length &gt; 1 ? join_path.apply( {}, arguments ) : arguments[0];","","      try {","        return fs.realpathSync( _path );","      } catch ( error ) {","","        try {","          return require.resolve( _path );","        } catch( error ) {","          return null;","        }","      }","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  console: {","    value: {","      /**","       * Prety Print complex objects","       *","       * @param data","       * @param depth","       *","       * @requires lodash","       * @method console_json","       * @return {Object} The first argument.","       */","      json: function json( data, depth ) {","","        try {","","          if( 'string' === typeof data &amp;&amp; ( 'object' === typeof depth || 'function' === typeof depth ) ) {","            console.log( \"\\n\" + '===============' + data + ' Properties===============' );","            data = depth;","            depth = arguments[2] || 1","          }","","          var output = require( 'util' ).inspect( data, false, 'number' === typeof depth ? ( depth - 1 ) : 1, true );","","          console.log( output );","","        } catch( error ) { console.error( 'Error with console.json()', error ); }","","        return arguments[0];","","      },","      /**","       * Pretty Print and Object's Methods","       *","       * @param data","       * @param object","       *","       * @requires lodash","       * @method console_methods","       * @return {Object} The first argument.","       */","      method: function method( data, object ) {","","        try {","","          if( 'string' === typeof data &amp;&amp; ( 'object' === typeof object || 'function' === typeof object ) ) {","            console.log( \"\\n\" + '===============' + data + ' Methods===============' );","            data = object;","          }","","          module.exports.json( _.methods( data ) );","","        } catch( error ) { console.error( 'Error with console.methods()', error ); }","","        return arguments[0];","","      },","      /**","       * Get all Object Keys","       *","       * @param data","       *","       * @method console_keys","       * @return {Object} The first argument.","       */","      keys: function keys( data ) {","","        function Iterate( target ) {","          var result = [];","          for( var key in target ) {","            result.push( key );","          }","          return result;","        }","","        console.log([ Object.keys( data ) + Object.getOwnPropertyNames( data ) + Iterate( data ) ]);","        return data;","      },","      /**","       * Draw data in table style","       *","       * @example","       *","       *    //console.table","       *    console.table([","       *      { .. row .. }","       *    ]);","       *","       *    //console.table","       *    console.table({","       *      'head': [ 'I/O ID', 'Endpoint', 'Description' ],","       *      'colWidths': [ 30, 50, 50 ],","       *      'rows': [","       *        [ 'one', 'two', 'three' ],","       *        [ 'one', 'two', 'three' ]","       *      ]","       *    });","       *","       * @param {object} data","       * @method table","       * @for Utility","       */","      table: function table( data ) {","","          try {","            var table = require( 'cli-table' );","          } catch (error) {","            return console.log( data );","          }","","          var rows = [];","","          if( _.isObject( data ) &amp;&amp; data.rows ) {","","            table = new table({","              'head': data.head || [],","              'colWidths': data.colWidths || []","            });","","            rows = data.rows;","","          } else {","            table = new table;","            rows = data;","          }","","          _.each( rows, function( row ) {","            if( !_.isEmpty( row ) ) { table.push( row ); }","          });","","          try {","            console.log( table.toString() );","          } catch( error ) {","            //console.trace( error.message, error );","          }","","        },","      /**","       * Console a variable report","       *","       * @method report","       * @for Utility","       * @param {object} arguments*","       */","      report: function report() {","          module.exports.console.json.apply( module.exports, arguments );","          module.exports.console.methods.apply( module.exports, arguments );","        }","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  merge: {","    /**","     * Merge object b with object a.","     *","     * @example","     *     var a = { foo: 'bar' }","     *       , b = { bar: 'baz' };","     *","     *     utils.merge(a, b);","     *     // =&gt; { foo: 'bar', bar: 'baz' }","     *","     * @param {Object} a","     * @param {Object} b","     * @return {Object}","     *","     * @source connect","     */","    value: function merge( a, b ){","","      if (a &amp;&amp; b) {","        for (var key in b) { a[key] = b[key]; }","      }","","      return a;","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  embed: {","    /**","     * Shallow Merge of Current Object with Another, and Function Mounting","     *","     * Functions in exact same manner as","     * the Connect utilities \"merge\" method","     *","     * @param {Object} target","     *","     * @method embed","     *","     * @chainable","     *","     * @return {Object}","     */","    value: function embed( target ) {","","      if( 'object' === typeof target || target.keys ) {","        for( var key in target ) {","          this[key] = target[ key ];","        }","      }","","      // If second argument is a function, we do some magic","      if( 'function' === typeof target ) {","        return Utility.embed( this );","      }","","      return this;","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  regroup: {","    /**","     * Re-group an object by a key, and optionally pick specific keys in the new object.","     *","     * If an object is passed, it is assumed to be grouped, thus flattening the values.","     * Arrays are used as they are.","     *","     * @param target","     */","    value: function regroup( object, group_key, keys ) {","","      if( _.isObject( object ) ) {","        object = _.flatten( _.values( object ) );","      }","","      function Group( item ) {","        return item[ group_key ];","      }","","      if( group_key ) {","        object = _.groupBy( object, Group );","      }","","      if( !_.isEmpty( keys ) ) {","","        keys = _.isArray( keys ) ? keys : [ keys ];","","        _.each( object, function( items, key ) {","          object[ key ] = _.map( items, function( item ) { return _.pick( item, keys ); })","        })","","      }","","      return object;","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  noop: {","    /**","     * Not a function, that's for sure.","     * This method does not accept any arguments.","     */","    value: function noop() {","      // console.log( arguments )","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  apply: {","    /**","     * Creates a continuation function with some arguments already applied.","     *","     * @uses async","     */","    value: require( 'async' ).apply,","    configurable: true,","    enumerable: true,","    writable: true","  },","  defaults: {","    /**","     * Configure Defaults for an Object","     *","     * @returns {Object}","     */","    value: function defaults( target, defaults ) {","","      // If no arguments, return empty object","      if( !target &amp;&amp; !defaults ) {","        return {};","      }","","      // Ensure target is an object","      target = Object.keys( target || {} ).length ? target : {};","","      // Lodash-it","      return require( 'lodash' ).defaults.apply( {}, arguments );","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  watch: {","    /**","     * Watch a Single Property","     *","     * @param {Object} prop","     * @param {callback} handler","     *","     * @method watch","     * @for abstract-utility","     *","     * @chainable","     *","     * @return {Object} newval","     */","    value: function watch( target, prop, handler ) {","      var oldval = target[prop];","      var newval = oldval;","      var description = Object.getOwnPropertyDescriptor( target, prop ) || {};","","      switch( typeof target[prop] ) {","        case 'function':break;","        case 'object':break;","        case 'string':break;","        case 'number':break;","      }","","      // Delete original property and replace with getter/setter if possible","      if( description.configurable === false || !( delete target[ prop ] ) ) {","        return target;","      }","","      return Object.defineProperty( target, prop, {","        get: function get() {","          return newval;","        },","        set: function set( val ) {","          oldval = newval;","","          // Set value","          newval = val;","","          // Execute hander and potentially mody value","          return newval = handler.call( target, prop, oldval, val );","","        },","        enumerable: description.enumerable,","        configurable: true","      });","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  unwatch: {","    /**","     * Unwatch a Single Property","     *","     * @param {Object} prop","     *","     * @method unwatch","     * @for abstract-utility","     *","     * @return null","     */","    value: function unwatch( prop ) {","      var val = this[prop];","      var description = Object.getOwnPropertyDescriptor( this, prop );","","      // Shouldn't happen, but just in case.","      if( description.configurable === false || !( delete this[ prop ] ) ) {","        return this;","      }","","      this[prop] = val;","","      return this;","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  query: {","    /**","     * Get property value using a dot notation path.","     *","     * @param obj","     * @param str","     * @returns {*}","     */","    value: function query( obj, str ) {","","      if( !str ) {","        return obj;","      }","","      if( 'object' === typeof str ) {","        // @todo Convert Object to dot notation, using the first full dot notation path.","      }","","      try {","        return str.split( '.' ).reduce( function( o, x ) {","          return o[x]","        }, obj);","","      } catch( error ) {","        return null;","      }","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  unwrap: {","    /**","     * Unwrap dot notation string to nested Object","     *","     * @example","     *","     *    Utility.unwrap( 'first_name', 'john' );  // -&gt; { first_name: 'John' }","     *    Utility.unwrap( 'name.first', 'john' );  // -&gt; { name: { first: 'John' } }","     *    Utility.unwrap( 'first_name' );          // -&gt; null","     *    Utility.unwrap( 'name.first' );          // -&gt; null","     *","     *    // Will not modify object if already exists","     *    Utility.unwrap( 'name.first.fail', 'Anything' );","     *","     * @param string","     * @param value","     * @param hash","     * @param seperator","     * @,etjpd unwrap","     * @return {*}","     */","    value: function unwrap( string, value, hash, seperator ) {","      if( hash == null ) { hash = {}; }","      if( seperator == null ) { seperator = '.'; }","      var parts = string.split( seperator );","      var refHash = hash;","      var depth = 0;","","      parts.forEach( function( part ) {","        if( depth == parts.length - 1 ) {","          refHash[part] = value;","        } else {","          if( refHash[part] == null ) { refHash[part] = {}; }","          refHash = refHash[part];","        }","","        depth++;","","      });","","      return hash;","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  mixin: {","    /**","     * Mixin current prototype into target object","     *","     * Honors property descriptor settings, if available.","     *","     * @param source","     * @param target","     */","    value: function mixin( source, target ) {","      source = this.prototype || arguments[0] || {};","      target = arguments.length == 2 ? arguments[1] : arguments[1] || {};","","      for( var key in source ) {","","        try {","          Object.defineProperty( target, key, Object.getOwnPropertyDescriptor( source, key ) || {","            value: source[key],","            enumerable: false,","            writable: true,","            configurable: true","          });","        } catch( error ) { console.error( error ); }","","      }","","      return target;","","    },","    configurable: true,","    writable: true,","    enumerable: false","  },","  inherit: {","    /**","     * Carefully Inherit Properties","     *","     * @param target {Object} Object to add properties to.","     * @param target {Object} Source object.","     * @returns {Object} Extended target.","     */","    value: function inherit( target, source ) {","      target = target || {};","      source = source || {};","","      //if( target instanceof source ) {}","","      Object.getOwnPropertyNames( source ).forEach( function( key ) {","        if( !target.hasOwnProperty( key ) ) { Object.defineProperty( target, key, Object.getOwnPropertyDescriptor( source, key ) ); }","      });","","      return target;","","    },","    enumerable: true","  },","  flatten: {","    /**","     * Flatten Array","     *","     * @todo Notice - nested arrays will not honor delimiter.","     */","    value: function flatten( data, options ) {","","      options = Utility.defaults( options, {","        delimiter: 1","      });","","      return Array.prototype.concat( data ).join( options.delimiter ).toLowerCase();","","","    },","    enumerable: true","  },","  flatten_obj: {","    value: function flatten_obj( obj ) {","","      var list = {};","","      (function callee(o, r) {","","        r = r || '';","","        if (typeof o != 'object') {","          return true;","        }","","        for (var c in o) {","","          if ( callee( o[c], r + \".\" + c ) ) {","","            list[r.substring(1) + \".\" + c] = o[c]","          }","        }","","        return false;","","      })(obj);","","      return list;","    },","    enumerable: true","  },","  inherit_full: {","    /**","     * Inherit the prototype methods from one constructor into another.","     *","     * Copy of the Node.js util.inherits method.","     *","     * @param {function} target Constructor function which needs to inherit the prototype.","     * @param {function} constructor Constructor function to inherit prototype from.","     */","    value: function inherit_prototype( target, constructor ) {","","      target.super_ = constructor;","","      target.prototype = Object.create( constructor.prototype, {","        constructor: {","          value: constructor,","          enumerable: false,","          writable: true,","          configurable: true","        }","      });","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  constructors: {","    /**","     * Walk up the prototype chain, creating a JSON-Schema-esque structure","     *","     * @param context {Object} The object to walk through.","     * @return {Object} JSON Schema-esque constructor chain.","     */","    value: function constructors( context, options ) {","      context = context || {};","      options = options || {};","","      var matches = {};","      var path = [];","","      do {","","        if( context.constructor ) {","          path.push( context.constructor.name );","","          Object.defineProperty( matches, context.constructor.name, {","            enumerable: true,","            writable: true,","            value: {","              name: context.constructor.name,","              properties: {","                constructor: {","                  type: typeof context.constructor,","                  properties: Object.getOwnPropertyNames( context.constructor )","                },","                prototype: {","                  type: typeof context.constructor.prototype,","                  properties: Object.keys( context.constructor.prototype )","                }","              }","            }","          });","","        }","","      } while ( context = Object.getPrototypeOf( context ) );","","      if( options.format === 'schema' ) {","        return matches || {};","      }","","      if( options.delimiter ) {","        return path.join( options.delimiter );","      }","","      return path;","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  unique: {","    /**","     * Return unique values of an array","     *","     * @param array","     * @returns {Array}","     */","    value: function unique( array ){","      var b = [];","      for(var i=0; i&lt;array.length; i++){","        if(b.indexOf(array[i]) == -1) b.push(array[i]);","      }","      return b;","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  hash: {","    /**","     * Generate a unique hash for an Object, using md5 on default.","     *","     * @example","     *","     *    var _hash = Utility.hash({ type: 'some_object', name: 'Bob'});","     *","     *    console.log( _hash ); // -&gt; 147ce3e2ccb7db6b928b303ce42bdafa","     *","     * @param obj {Object} Object to generate a hash for.","     * @param options {Object} options for hash generation.","     * @param options {String} options.type Type of hash to generate, defaulting to md5.","     * @param options {String} options.silent_fail Do not throw errors, return empty string if there was an error when true.","     *","     */","    value: function hash( obj ) {","","      var result;","      var string = JSON.stringify( arguments[0] || {} );","      var options = Utility.defaults( arguments[1], {","        type: 'md5',","        silent_fail: true,","        random: false","      });","","      try {","","        if( options.random ) {","          string = Math.random().toString();","        }","","        result = require( 'crypto' ).createHash( options.type.toLowerCase() ).update( string ).digest( 'hex' );","      } catch ( error ) {","        result = options.silent_fail ? '' : error;","      }","","      return result;","","    },","    configurable: true,","    enumerable: true,","    writable: true","  },","  auto: { value: require( 'async' ).auto },","  queue: { value: require( 'async' ).queue },","  times: { value: require( 'async' ).times },","  extend: { value: require( 'lodash' ).extend },","  values: { value: require( 'lodash' ).values },","  map: { value: require( 'lodash' ).map },","  each: { value: require( 'lodash' ).each },","  reduce: {","    value: function reduce(arr, iterator, memo) {","","      if (arr.reduce) {","        return arr.reduce(iterator, memo);","      }","","      _each(arr, function (x, i, a) {","        memo = iterator(memo, x, i, a);","      });","","      return memo;","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  keys: {","    /**","     *","     * @source async","     */","    value: function keys(obj) {","","      if (Object.keys) {","        return Object.keys(obj);","      }","","      var keys = [];","      for (var k in obj) {","        if (obj.hasOwnProperty(k)) {","          keys.push(k);","        }","      }","      return keys;","","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  toArray: { value: require( 'lodash' ).toArray },","  where: { value: require( 'lodash' ).where },","  if: {","    value: {","      \"PlainObject\": require( 'lodash' ).isPlainObject,","      \"Function\": require( 'lodash' ).isFunction,","      \"Object\": require( 'lodash' ).isObject,","      \"String\": require( 'lodash' ).isString,","    },","    configurable: false,","    enumerable: true,","    writable: true","  },","  log: {","    /**","     * Output variables to console log.","     *","     * Settings can be configured like so:","     *    abstract.log.config.depth = 2;","     *    abstract.log.config.colors = true;","     *","     * @param {Object|String} data A config object","     * @example","     *      Abstract.log( data )","     *","     * method log","     * @return {Object} Abstract constructor.","     * @chainable","     */","    value: Object.defineProperties( function() {","","      var output = {};","","      // Capture Event Name","      if( this.event ) {","        output.event = this.event;","      }","","      Utility.each( arguments, function( item, key ) {","","        if( Utility.if.PlainObject( item ) ) {}","","        output[ 'format' ] = typeof item;","","        if( key === 0 ) {","          output[ 'data' ] = item;","        } else {","          output[ key ] = item;","        }","","      });","","      console.log( require( 'util' ).inspect( output, ( Utility ).log.config ) );","","      return this;","    }, { \"config\": { value: { \"showHidden\": true, \"depth\": 2, \"colors\": true }, \"writable\": true } }),","    enumerable: true","  }","});"];
